### PR TITLE
Favor HaltExceptions, improve implicit halting, take out unnecessary code, simplify, and much more.

### PR DESCRIPTION
**Pull request by 2rs2ts**
The goal of this PR was to encourage the use of `HaltException`s, and the methods to create them more easily, in downstream applications. Some other cleanup was done since it was convenient or mildly related.

## Additions

1. Added documentation to some methods.

## Removals

1. `sendErrorMapResponse` – this is completely unused by us.
2. `toJsonErrorsMap` – logically follows that it should be removed if `sendErrorMapResponse` is.
3. `toJsonErrors` – useless alias for `toJsonBody`.
4. `ValueConstants` – it only contained an alias for the UTF-8 charset, and `nio` already has a nice constant for this that should be used.
5. `HttpUtil.UTF_8` – same reason as above.
6. log message in `handleUnexpectedRequestError` which could leak sensitive information to the log file.

## Modifications

1. Error responses do not show exception messages. (The implicit classes which provide halting methods still use exception messages by default, to my chagrin, but I'm not sure I want to change it at this time.)
2. You do not tell `HttpResponse`s to the `HttpResourceActor` anymore. Use `complete`, `completeToJson`, and the relevant `sendError` methods.
3. `handleHttpResponse` is now package-private as a result of the above change.
4. `handleErrors` is now called `createErrorResponse`, which describes what it does better.
5. `createErrorResponse` adds the language header instead of delegating that to the `AbstractResourceActor`.
6. Implicit class for `Either[Throwable, A]` which provides halting methods returns `Try`s instead of `Future`s because you can easily convert a `Try` to a `Future` but not the other way around.

### CODE REVIEW

**Comment by taylorleese** [comment]
@2rs2ts Sadly, I've never really liked ```HaltException```. I still don't really understand why we need it.

**Comment by taylorleese** [comment]
@2rs2ts It seems like ```HaltException``` is a generic catch all for whenever we want to stop the http request pipeline. Do we gain something with it over specific exceptions for specific failures?

**Comment by taylorleese** [comment]
@2rs2ts Is number 6 in modifications still valid? I couldn't find it.

**Comment by 2rs2ts** [comment]
@taylorleese Let me address your points in two parts:

First, I think `HaltException` is handy in the way you describe. It's a good way for any flow to generically stop, as it gives you everything you need to respond. I didn't want to overreach in this PR, but I'd actually like to merge `AbstractResourceActor` into `HttpResourceActor` and have clients inherit from that, and to reinstate it handling `Status.Failure` cases before clients do, but only matching on ones containing `HaltException`s. People should be able to write flows that look like this:

```scala
// get some data from some service, which throws an exception if it's not found
Try(getSomething(x)).orHaltWithMessage(NotFound)(t => s"${t.getClass.getSimpleName}: Could not find $x") match {
  case Success(result) => complete(HttpResponse(OK, result.toJsonBody))
  case Failure(haltEx) => sendError(haltEx)
}
```

Of course, they should also be able to define more complex flows for custom exceptions. But `HaltException` technically has everything you need. I'm also considering replacing the `Exception => HttpResponse` flow with an `Exception => HaltException` flow. Then clients would, instead of creating `HttpResponses` in `createErrorResponse`, they would create `HaltExceptions` and the original `createErrorResponse` flow becomes a way to guarantee that certain headers are always set and other goodies.

Secondly, here's the line for modification 6: https://github.com/paypal/cascade/pull/95/files#diff-8feb134007befd7bc880eab05ee2c618R199

**Comment by taylorleese** [comment]
@2rs2ts Alright, if you think it provides value. It's been there a long time already anyways so I'm not going to stress about it staying around. In regards to combining the resource and http actors, I know @milesoc preferred them to stay distinct last time I discussed it with him. You should chat with him about it.

**Comment by 2rs2ts** [comment]
I will. I think that this code has changed enough to the point where so much of `HttpResourceActor` depends on `AbstractResourceActor` doing the right thing that it's a violation of inheritance principles where the parent has to know about the child implementation.

**Comment by devmage** [comment]
@2rs2ts that's definitely an observation worth discussing. Do you see that because of the relatively thin code in `HttpResourceActor`?

File: common/src/main/scala/com/paypal/cascade/common/string/string.scala [hunk 1]
```diff
@@ -33,7 +33,7 @@ package object string {
      * @return an array of bytes from UTF-8 encoding
      */
     def getBytesUTF8: Array[Byte] = {
-      inner.getBytes(charsetUtf8)
+      inner.getBytes(UTF_8)
```
Comment (devmage) [inline common/src/main/scala/com/paypal/cascade/common/string/string.scala]: Good to know this constant exists!

**Comment by devmage** [comment]
@2rs2ts I agree with the utility of `HaltException`, and I generally like what I see in this PR. These classes were due for some cleanup, so thank you for taking this step.

As for merging the two actors as you propose, I'd want to get more input from others before signing on to that idea. However, that discussion is also outside the scope of this PR, so we'll talk about it when it comes up.

For this PR, :shipit: when you're ready.

**Pull request merged**